lib/transport/transport-udp-socket.c

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (c) 2002-2019 Balabit
+ * Copyright (c) 1998-2019 Balázs Scheidler
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+#include "transport-udp-socket.h"
+#include "transport/transport-socket.h"
+#include "scratch-buffers.h"
+#include "str-format.h"
+
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <errno.h>
+
+typedef struct _LogTransportUDP LogTransportUDP;
+struct _LogTransportUDP
+{
+  LogTransportSocket super;
+  GSockAddr *local_addr;
+};
+
+#if defined(SYSLOG_NG_HAVE_CTRLBUF_IN_MSGHDR)
+static void
+_feed_aux_from_cmsg(LogTransportUDP *self, LogTransportAuxData *aux, struct msghdr *msg)
```

Comment:
I am curious if this has a performance impact? From now for every udp packet, we need to read the ancillary data too. Can this cause problem? Should we put the code behind optionality?

Reply:
I don't think it would be a huge issue (well, we already incur a lot of penalty by allocating a GSockAddr instance for every datagram, now we do two). The ancillary data is retrieved in the same syscall and should be much smaller than the datagram itself.

What I would do if UDP performance should be improved is to embed the GSockAddr instance into LogMessage and avoid the memory allocation altogether, possibly doing a special case for ipv4/ipv6 traffic.

Reply:
let me know if you have a different opinion.

Reply:
I'm afraid I cannot tell. The syscall potentially lasts longer, and processing the result is also involved. The safest is to do some measurement. I will ask help from @pzoleex, and try to validate this.

Reply:
In the meantime I've started working on containing the GSockAddr instances right within LogMessage, but that's a huge change, which I'd prefer do in a separate PR. Initial performance testing suggests that there's ~5% improvement in performance if we do that, assuming that _both_ addresses are stored, which was around 350k/sec per thread on my laptop (without the inlineing), and ~365k (with the inlining).

What this boils down to is that even storing two instances instead of one is not that much of a headache (e.g. 2-3% max), and that only applies to UDP reception speed.

To improve our UDP story, I'd do something like the recvmmsg() patches in #2755 and also do the inlining. That should give us a considerable boost.

With all this said, I think merging this should be fine for now from the performance standpoint. GSockAddr inlining, UDP performance boost could come later.

Reply:
I disagree with that. I am happy to hear there are other ways to improve udp performance. But if this patch introduces a performance problem, we need to know it now. Because if there is a performance problem, someone needs to come back at some point to fix it, no matter the other performance improvements in other parts of the code.

The action point is at the testing team now. And I am really sorry that the performance takes this long, we are just before release and there is no testing capacity available. I hope we can have progress next week.

Reply:
@pzoleex did us the measurement, and you were right: there is no performance problem. Sorry for the delay! Marking as resolved. 

```
nightly
100k eps -> Average outgoing lines/sec: 88130 msg/sec
200k eps -> Average outgoing lines/sec: 104996 msg/sec
new code:
100k eps -> Average outgoing lines/sec: 86271 msg/sec
200k eps -> Average outgoing lines/sec: 103333 msg/sec
original code:
100k eps -> Average outgoing lines/sec: 88066 msg/sec
200k eps -> Average outgoing lines/sec: 105556 msg/sec
```
Measurement: simple udp source -> file destination on single thread. EPS measured on the output file.